File: src/renamer.cpp
#include "renamer.hpp"
#include <algorithm>
#include <cctype>
#include <fmt/compile.h>
#include <fmt/core.h>
#include <fmt/format.h>
#include <iterator>
#include <stdexcept>
#include <unordered_map>
#include <vector>

using namespace fmt::literals;

std::optional<std::string> check_format(std::string const& format, fmt::format_args args) {
    try {
        if (fmt::vformat(format, args) == format)
            return "Formatted cannot be the same as the input. (Missing {}?)";
    } catch (const fmt::format_error& error) {
        return std::make_optional<std::string>(error.what());
    }
    return std::nullopt;
}

Fmt::Fmt() {
    std::unordered_map<std::string, const StringFmtBase*> formats = {
        { "classes", &classes },
        { "consts", &consts },
        { "functions", &functions },
        { "names", &names },
        { "vars", &vars },
        { "methods", &methods },
        { "errors", &errors },
        { "unknown_clientbound_packet", &unknown_clientbound_packet },
        { "packet_subhandler", &packet_subhandler },
        { "clientbound_packet", &clientbound_packet },
        { "serverbound_packet", &serverbound_packet },
    };
    for (const auto& [name, fmt] : formats) {
        auto error = fmt->valid();
        if (error) {
            throw std::runtime_error(fmt::format(
                FMT_COMPILE("\"{value}\" is not a valid format string for [{name}]: {reason}"),
                "value"_a  = fmt->value,
                "name"_a   = name,
                "reason"_a = *error));
        }
    }
}

void Fmt::from_json(json& config, utils::Logger& logger) {
    if (!config.is_object() || !config.contains("formats"))
        return;

    for (auto& [name, field] : get_fields()) {
        json::json_pointer ptr(name);
        if (!config.contains(ptr))
            continue;

        auto& child = config[ptr];
        if (!child.is_string()) {
            std::replace(name.begin(), name.end(), '/', '.');
            logger.error("'{}' must be a string.\n", name.substr(1));
            continue;
        }

        auto value = child.get<std::string>();
        if (value.empty())
            continue;

        auto error = field->valid(value);
        if (error) {
            std::replace(name.begin(), name.end(), '/', '.');
            logger.error("'{}' is not a valid format: {}\n", name.substr(1), *error);
        } else {
            field->value = value;
        }
    }
}

json Fmt::to_json() {
    Fmt fmt;
    json config;

    for (auto& [name, field] : fmt.get_fields())
        config[name] = field->value;

    return config.unflatten();
}

Fmt::Fields Fmt::get_fields() {
    return {
        { "/formats/typenames/classes", &classes },
        { "/formats/typenames/consts", &consts },
        { "/formats/typenames/functions", &functions },
        { "/formats/typenames/names", &names },
        { "/formats/typenames/vars", &vars },
        { "/formats/typenames/methods", &methods },
        { "/formats/typenames/errors", &errors },
        { "/formats/packets/packet_subhandler", &packet_subhandler },
        { "/formats/packets/unknown_clientbound_packet", &unknown_clientbound_packet },
        { "/formats/packets/clientbound", &clientbound_packet },
        { "/formats/packets/serverbound", &serverbound_packet },
        { "/formats/packets/tribulle/clientbound", &tribulle_clientbound_packet },
        { "/formats/packets/tribulle/serverbound", &tribulle_serverbound_packet },
    };
}

Renamer::Renamer(std::shared_ptr<abc::AbcFile> const& abc, Fmt const& fmt) : abc(abc), fmt(fmt) { }

bool Renamer::invalid(std::string const& name) { return invalid(std::string_view(name)); }
bool Renamer::invalid(std::string_view name) {
    return std::any_of(
        std::begin(name), std::end(name), [](const char c) { return !std::isprint(c); });
}

void Renamer::rename() {
    for (auto& klass : abc->classes)
        rename(klass);

    for (auto& method : abc->methods)
        rename(method);
}

void Renamer::rename(abc::Class& klass) {
    if (klass.name != 0 && invalid(klass.get_name()))
        klass.rename(fmt.classes.format(++counters.classes));

    if (klass.super_name != 0 && invalid(klass.get_super_name()))
        klass.rename_super(fmt.classes.format(++counters.classes));

    for (auto& trait : klass.ctraits)
        rename(trait);

    for (auto& trait : klass.itraits)
        rename(trait);
}
void Renamer::rename(abc::Trait& trait) {
    if (trait.name != 0 && invalid(trait.get_name())) {
        switch (trait.kind) {
        case swf::abc::TraitKind::Const:
            trait.rename(fmt.consts.format(++counters.consts));
            break;
        case swf::abc::TraitKind::Method:
            trait.rename(fmt.methods.format(++counters.methods));
            break;
        case swf::abc::TraitKind::Function:
            trait.rename(fmt.functions.format(++counters.functions));
            break;
        default:
            trait.rename(fmt.names.format(++counters.names));
            break;
        }
    }
}

void Renamer::rename(abc::Method& method) {
    int counter = 0;
    if (method.exceptions.size() == 1) {
        rename(method.exceptions[0]);
    } else {
        for (auto& err : method.exceptions)
            rename(err, ++counter);
    }

    for (auto& trait : method.traits)
        rename(trait);
}
void Renamer::rename(abc::Exception& err) {
    if (invalid(err.get_var_name()))
        err.rename_var_name("error");
}
void Renamer::rename(abc::Exception& err, int counter) {
    if (invalid(err.get_var_name())) {
        auto name = fmt.errors.format(counter);
        err.rename_var_name(name);
    }
}

File: src/renamer.rs
use crate::{
    fmt,
    rename::{PoolRenamer, Rename, RenameSuper},
};
use anyhow::Result;
use rabc::{
    abc::{Class, ConstantPool, Exception, Method, Trait},
    Abc,
};

macro_rules! counter {
    ($name:ident) => {
        fn $name(&mut self) -> u32 {
            let value = self.$name;
            self.$name += 1;
            value
        }
    };
}

#[derive(Debug, Default)]
struct Counters {
    classes: u32,
    consts: u32,
    functions: u32,
    _names: u32,
    vars: u32,
    methods: u32,
}
impl Counters {
    counter!(classes);
    counter!(consts);
    counter!(functions);
    counter!(_names);
    counter!(vars);
    counter!(methods);
}

#[derive(Debug, Default)]
pub struct Renamer {
    counters: Counters,
}

impl Renamer {
    pub fn invalid(name: &str) -> bool {
        name.chars().any(|c| !c.is_alphabetic())
    }
    fn invalid_multiname(cpool: &ConstantPool, index: u32) -> Result<bool> {
        Ok(Self::invalid(cpool.get_str(cpool.get_mn_name(index)?)?))
    }

    pub fn rename_all(&mut self, abc: &mut Abc, cpool: &mut ConstantPool) -> Result<()> {
        for class in &abc.classes {
            self.rename_invalid_class(cpool, class)?;
        }

        for method in &abc.methods {
            self.rename_invalid_method(cpool, method)?;
        }
        Ok(())
    }
    fn rename_invalid_class(&mut self, cpool: &mut ConstantPool, class: &Class) -> Result<()> {
        if class.name != 0 && Self::invalid_multiname(cpool, class.name)? {
            class.rename(cpool, fmt::classes(self.counters.classes()))?;
        }
        if class.super_name != 0 && Self::invalid_multiname(cpool, class.super_name)? {
            class.rename_super(cpool, fmt::classes(self.counters.classes()))?;
        }

        for cls_trait in class.ctraits.iter().chain(class.itraits.iter()) {
            self.rename_invalid_trait(cpool, cls_trait)?;
        }
        Ok(())
    }

    fn rename_invalid_trait(&mut self, cpool: &mut ConstantPool, cls_trait: &Trait) -> Result<()> {
        let name = cls_trait.name();
        if name != 0 && Self::invalid_multiname(cpool, name)? {
            let name = match cls_trait {
                Trait::Const(_) => fmt::consts(self.counters.consts()),
                Trait::Method(_) => fmt::methods(self.counters.methods()),
                Trait::Function(_) => fmt::functions(self.counters.functions()),
                _ => fmt::vars(self.counters.vars()),
            };
            cls_trait.rename(cpool, name)?;
        }
        Ok(())
    }

    fn rename_invalid_method(&mut self, cpool: &mut ConstantPool, method: &Method) -> Result<()> {
        for (i, err) in method.exceptions.iter().enumerate() {
            Self::rename_invalid_exception(cpool, err, i as u32)?;
        }

        for mtrait in &method.traits {
            self.rename_invalid_trait(cpool, mtrait)?;
        }
        Ok(())
    }

    fn rename_invalid_exception(
        cpool: &mut ConstantPool,
        exception: &Exception,
        counter: u32,
    ) -> Result<()> {
        if Self::invalid_multiname(cpool, exception.var_name)? {
            exception.rename(cpool, fmt::errors(counter))?;
        }
        Ok(())
    }
}
